toponetx/classes/path_complex.py

```diff
@@ -0,0 +1,657 @@
+    allowed_paths : Iterable[tuple[Hashable]], optional
+        An iterable of allowed boundaries. If None, only obvious boundaries are constructed (sub-sequences where the first or the last index is omitted).
+    max_rank : int, default=3
+        The maximal length of a path in the path complex.
+    attr: keyword arguments, optional
+        Additional attributes to be associated with the path complex.
+    """
+
+    def __init__(
+        self,
+        paths: nx.Graph | Iterable[Sequence[Hashable]] = None,
+        name: str = "",
+        reserve_sequence_order: bool = False,
+        allowed_paths: Iterable[tuple[Hashable]] = None,
+        max_rank: int = 3,
+        **kwargs,
+    ) -> None:
+
+        super().__init__(name=name, **kwargs)
+
+        self._path_set = PathView()
+        self._reserve_sequence_order = reserve_sequence_order
+        if allowed_paths is not None:
+            self._allowed_paths = set(allowed_paths)
+        else:
+            self._allowed_paths = set()
+
+        if isinstance(paths, nx.Graph):
+            # compute allowed_paths in order to construct boundary incidence matrix/adj matrix.
+            if self._allowed_paths is not None:
+                self._allowed_paths = self.compute_allowed_paths(
+                    paths,
+                    reserve_sequence_order=reserve_sequence_order,
+                    max_rank=max_rank,
+                )
+
+            # get feature of nodes and edges if available
+            for path, data in paths.nodes(data=True):
+                self.add_path(path, **data)
+            for u, v, data in paths.edges(
+                data=True
+            ):  # so far, path complex only supports undirected graph
+                if (str(u) > str(v)) and not reserve_sequence_order:
+                    u, v = v, u
+                self.add_path((u, v), **data)
+
+            # add all simple paths
+            self.add_paths_from(self._allowed_paths)
+
+        elif isinstance(paths, list) or isinstance(paths, tuple):
+            if len(paths) > 0:
+                paths = [tuple(path) for path in paths]
+                self.add_paths_from(set(paths))
+        elif paths is not None:
+            raise TypeError(
+                "Input paths must be a graph or an iterable of paths as lists or tuples."
+            )
+
+    def add_paths_from(self, paths: Iterable[Sequence[Hashable] | Path]) -> None:
+        """
+        Add paths from an iterable of paths.
+
+        Parameters
+        ----------
+        paths : Iterable[Sequence[Hashable] or Path]
+            an iterable of paths as lists, tuples, or Path objects.
+        """
+        if isinstance(paths, Hashable):
+            raise TypeError("Paths must be a iterable of paths as lists, tuples.")
+        paths_clone = paths.copy()
+        for p in paths_clone:
+            self.add_path(p)
+
+    def add_path(self, path: Hashable | Sequence[Hashable] | Path, **attr) -> None:
```

Comment:
define a  path mathmetically. Remember that the user does not have to know path complex very well to be able to use this method so an easy example should guide the user to create a path without much effort and hence utilize the complex also without much effort.

Reply:
Just added.